module/dkg/client.go

```diff
@@ -26,43 +27,51 @@ import (
 // Client is a client to the Flow DKG contract. Allows functionality to Broadcast,
 // read a Broadcast and submit the final result of the DKG protocol
 type Client struct {
-	dkgContractAddress string
-	accountAddress     sdk.Address
-	accountKeyIndex    uint
-	flowClient         module.SDKClientWrapper
-	signer             sdkcrypto.Signer
+	epochs.BaseContractClient
 
-	env templates.Environment // the required contract addresses to be used by the Flow SDK
+	log zerolog.Logger
+	env templates.Environment
 }
 
 // NewClient initializes a new client to the Flow DKG contract
-func NewClient(flowClient module.SDKClientWrapper, signer sdkcrypto.Signer, dkgContractAddress, accountAddress string, accountKeyIndex uint) *Client {
-	return &Client{
-		flowClient:         flowClient,
-		dkgContractAddress: dkgContractAddress,
-		signer:             signer,
-		accountKeyIndex:    accountKeyIndex,
-		accountAddress:     sdk.HexToAddress(accountAddress),
-		env:                templates.Environment{DkgAddress: dkgContractAddress},
+func NewClient(log zerolog.Logger,
+	flowClient module.SDKClientWrapper,
+	signer sdkcrypto.Signer,
+	dkgContractAddress,
+	accountAddress string,
+	accountKeyIndex uint) (*Client, error) {
+
+	base, err := epochs.NewBaseContractClient(log, flowClient, accountAddress, accountKeyIndex, signer, dkgContractAddress)
+	if err != nil {
+		return nil, err
 	}
+	env := templates.Environment{DkgAddress: dkgContractAddress}
+
+	return &Client{
+		BaseContractClient: *base,
+		env:                env,
+	}, nil
 }
 
 // Broadcast broadcasts a message to all other nodes participating in the
 // DKG. The message is broadcast by submitting a transaction to the DKG
 // smart contract. An error is returned if the transaction has failed.
 func (c *Client) Broadcast(msg model.BroadcastDKGMessage) error {
 
+	started := time.Now()
+
 	ctx, cancel := context.WithTimeout(context.Background(), epochs.TransactionSubmissionTimeout)
 	defer cancel()
 
 	// get account for given address
-	account, err := c.flowClient.GetAccount(ctx, c.accountAddress, grpc.EmptyCallOption{})
+	account, err := c.Client.GetAccount(ctx, c.Account.Address, grpc.EmptyCallOption{})
 	if err != nil {
 		return fmt.Errorf("could not get account details: %w", err)
 	}
+	c.Account = account
```

Comment:
If I understand correctly, this is to make sure the sequence number is in sync after requesting the account. Practically we are treating `c.Account` as a local variable here, since it is only valid immediately after a call to `GetAccount`. If we forget to call `GetAccount` and update `c.Account` before using it, we could have stale account data. Because of this, I would suggest using a local `account` variable and removing the `BaseContractClient.Account` field, since that field cannot be guaranteed to be up-to-date.
```go
account, err := c.Client.GetAccount()
// use the local account variable rather than `c.Account`
```

Reply:
Yeah the intention was to keep the `SequenceNumber` synced. But your suggestion is a better implementation. Thanks